### ISSUE TITLE
Perl locale disagrees with Linux sort

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8437]
Migrated from [rt.perl.org#39087](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39087) (status was 'rejected')

Searchable as RT39087$

**Comment by p5pRT** [comment]
### From felix@crowfix.com

#### Created by felix@crowfix.com



Run these two commands in a shell&#8203;:

&nbsp;    \(echo '/'; echo '?'\) | sort

&nbsp;    \(echo '/'; echo '?'\) | perl \-e 'print $\_ foreach sort \<>'

&nbsp;    \(echo '/'; echo '?'\) | perl \-e 'use locale;print $\_ foreach sort \<>'

and both print the expected

&nbsp;    /
&nbsp;    ?

as long as LC\_COLLATE is C\.  If you switch to en\_US\.utf\-8\, sort
reverses the two\.  Perl reverses them only if you "use locale;"\, and
that is as expected\.

Now try the two values "bin" and "/bin"\.  With locale "C"\, /bin comes
first under all three commands\.  With locale en\_US\.utf\-8\, sort puts
"bin" first\, but perl keeps them in "C" order regardless of "use
locale;"\.

I do not know which command is correct\, or what UTF\-8 collation order
is supposed to be\.  My hunch is that Perl is wrong because UTF seems
to consider '/' as some odd punctuation which goes at the end after
all real characters\, but that is about as scientific as any other
guess I might make\.

All I know is that perl and sort disagree\.  I have also tried this in
postgreSQL\, which is somewhat of a pain since a brand new test
database has to be created with the specific locale\, but it agrees
with sort\.

I am more than happy to do anything I can to clarify this or try
things for you\.  I don't even claim this to be a real bug; I may be
confused\, or sort may be wrong\.  I have been using Perl since
1990\-something and version 4\, and it's about time I contributed back a
little something\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.8:

Configured by Gentoo at Thu Apr 20 16:49:42 PDT 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.15-gentoo-r7, archname=x86_64-linux-thread-multi
    uname='linux crowfix 2.6.15-gentoo-r7 #1 smp preempt mon mar 6 08:48:06 pst 2006 x86_64 amd opteron(tm) processor 244 gnulinux '
    config_args='-des -Darchname=x86_64-linux-thread -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dcc=x86_64-pc-linux-gnu-gcc -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr -Dlocincpth=  -Doptimize=-O3 -march=k8 -mtune=k8 -pipe -fomit-frame-pointer -Duselargefiles -Dd_semctl_semun -Dscriptdir=/usr/bin -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dinstallman1dir=/usr/share/man/man1 -Dinstallman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3pm -Dinc_version_list=5.8.0 5.8.0/x86_64-linux-thread-multi 5.8.2 5.8.2/x86_64-linux-thread-multi 5.8.4 5.8.4/x86_64-linux-thread-multi 5.8.5 5.8.5/x86_64-linux-thread-multi 5.8.6 5.8.6/x86_64-linux-thread-multi 5.8.7 5.8.7/x86_64-linux-thread-multi  -Dcf_by=Gentoo -Ud_csh -Dusethreads -Di_ndbm -Di_gdbm -Di_db -Dusrinc=/usr/include/gentoo-multilib/amd64 -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=define uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='x86_64-pc-linux-gnu-gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3 -march=k8 -mtune=k8 -pipe -fomit-frame-pointer',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement'
    ccversion='', gccversion='3.4.6 (Gentoo 3.4.6-r1, ssp-3.4.5-1.0, pie-8.7.9)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='x86_64-pc-linux-gnu-gcc', ldflags =' -L/usr/local/lib64'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lpthread -lnsl -lndbm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib64'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/lib64/perl5/vendor_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib64/perl5/vendor_perl/5.8.8
    /usr/lib64/perl5/vendor_perl/5.8.5
    /usr/lib64/perl5/vendor_perl/5.8.6
    /usr/lib64/perl5/vendor_perl/5.8.7
    /usr/lib64/perl5/vendor_perl
    /usr/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib64/perl5/site_perl/5.8.8
    /usr/lib64/perl5/site_perl/5.8.4
    /usr/lib64/perl5/site_perl/5.8.5
    /usr/lib64/perl5/site_perl/5.8.7
    /usr/lib64/perl5/site_perl
    /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi
    /usr/lib64/perl5/5.8.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/felix
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=~/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/3.4.6:/opt/sourcenav/bin:/opt/blackdown-jdk-1.4.2.03/bin:/opt/blackdown-jdk-1.4.2.03/jre/bin:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/qt/2/bin:/usr/games/bin:/usr/share/omniORB/bin/scripts:/var/spool/fax/bin:/opt/insight/bin:/usr/share/karamba/bin:/var/qmail/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Fri\, 05 May 2006 12&#8203;:03&#8203;:59 \-0700\, "felix@&#8203;crowfix\.com \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> wrote

> \# New Ticket Created by  felix@&#8203;crowfix\.com 
> \# Please include the string&#8203;:  \[perl \#39087\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39087 >

> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Run these two commands in a shell&#8203;:
> 
>     \(echo '/'; echo '?'\) | sort
> 
>     \(echo '/'; echo '?'\) | perl \-e 'print $\_ foreach sort \<>'
> 
>     \(echo '/'; echo '?'\) | perl \-e 'use locale;print $\_ foreach sort \<>'
> 
> and both print the expected
> 
>     /
>     ?
> 
> as long as LC\_COLLATE is C\.  If you switch to en\_US\.utf\-8\, sort
> reverses the two\.  Perl reverses them only if you "use locale;"\, and
> that is as expected\.
> 
> Now try the two values "bin" and "/bin"\.  With locale "C"\, /bin comes
> first under all three commands\.  With locale en\_US\.utf\-8\, sort puts
> "bin" first\, but perl keeps them in "C" order regardless of "use
> locale;"\.

Perl's sort under the locale pragma relies on strxfrm\(\) for
the string comparison\.

cf\. perllocale\.pod&#8203;:

&nbsp;    USING LOCALES
&nbsp;    The use locale pragma
&nbsp;    \- The comparison operators \(lt\, le\, cmp\, ge\, and gt\) and
&nbsp;      the POSIX string collation functions strcoll\(\) and strxfrm\(\)
&nbsp;      use LC\_COLLATE\. sort\(\) is also affected if used without an
&nbsp;      explicit comparison function\, because it uses cmp by default\.

In other words\, Perl's sort \(and cmp also\) does just what strxfrm\(\) does\.

If your strxfrm\(\) thinks that '?' is before '/' and "/bin" is
before "bin"\, Perl's sort must obey it\.

Regards\,
SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'rejected'
